abstract_domains/numerical/octagon.py

```diff
@@ -0,0 +1,578 @@
+from copy import deepcopy
+from enum import Enum
+from functools import reduce
+
+from abstract_domains.lattice import BottomMixin
+from abstract_domains.numerical.dbm import IntegerCDBM
+from abstract_domains.numerical.interval import IntervalLattice, IntervalDomain
+from abstract_domains.numerical.numerical import NumericalMixin
+from abstract_domains.state import State
+from core.expressions import *
+from typing import List, Set, Tuple, Union
+from math import inf, isinf
+
+from abstract_domains.numerical.linear_forms import SingleVarLinearForm, LinearForm, InvalidFormError
+
+from core.expressions_tools import ExpressionVisitor, ExpressionTransformer, \
+    make_condition_not_free, simplify
+
+Sign = UnaryArithmeticOperation.Operator
+PLUS = Sign.Add
+MINUS = Sign.Sub
+
+
+def _index_shift(sign: Sign):
+    return 1 if sign == MINUS else 0
+
+
+class OctagonLattice(BottomMixin, NumericalMixin):
```

Comment:
This class desperately needs docstrings. I would also rename the methods that are not supposed to be used from the outside with an underscore in front.